Added Stochastic Frank-Wolfe variants + example + tests

GeoffNN opened the pull request:
Implemented Stochastic Frank-Wolfe from [our paper](https://arxiv.org/abs/2002.11860).

fabianp commented:
Nice start! Could you please add some tests and make sure it works when A is a spare matrix?

GeoffNN commented:
Will do!

GeoffNN commented:
Added tests, and changed API to match FW (lmo instead of constraint object as argument).
It works for sparse A matrices, and sparse x0 initial points.

fabianp commented:
Since some of these algorithms are very similar, would it make sense to have one sfw algorithm with a variant keyword that can take the options {"SAG" | "LF" | "MK"} or something along those lines? I think this would reduce some of the redundancies

GeoffNN commented:
Agreed, I’ll refactor.



On Mon 4 May 2020 at 21:40, Fabian Pedregosa <notifications@github.com>
wrote:

> Since some of these algorithms are very similar, would it make sense to
> have one sfw algorithm with a variant keyword that can take the options
> {"SAG" | "LF" | "MK"} or something along those lines? I think this would
> reduce some of the redundancies
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openopt/copt/pull/40#issuecomment-623664760>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2EQTKWVTUFH2LSCWJUIJLRP4K2ZANCNFSM4MYHKTEQ>
> .
>
-- 


*Geoffrey Négiar*
*PhD Candidate - UC Berkeley, EECS departmentX2013*

fabianp commented on copt/randomized.py:
```diff
@@ -691,3 +693,149 @@ def epoch_iteration_template(
+          the cause of the termination. See `scipy.optimize.OptimizeResult`
+          for a description of other attributes.
+
+    """
```
Please add the appropriate references, as in **minimize_frank_wolfe**

fabianp replied:
(for the SAG version reference our arxiv v2)

GeoffNN replied:
Done

fabianp commented:
Code looks good to me.

@gideonite do you have any comments?

fabianp commented:
@gideonite : do you have plans to review this or we should merge without your review?

gideonite commented:
Sorry for taking so long. Haven't had a chance to review it yet but will finally have time tomorrow. If there's urgency feel free to merge without me. @fabianp

fabianp commented:
No problem, I was just wondering if you were interested at all in reviewing. If so we can wait

gideonite commented on copt/randomized.py:
```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+def step_size_sfw(variant):
+    if variant in {'SAG', 'SAGA'} :
+        def step_sizes_SAG_A(t, n_samples=None):
+            step_size_x = 2 / (t+2)
```
`2. / (t+2)` just to be sure?

gideonite commented on copt/randomized.py:
```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+            return step_size_x, None
+        return step_sizes_SAG_A
+
+    if variant == 'MK':
```
Prefer "moktari" or something more. At the very least, define this somewhere in the file.

gideonite commented on copt/randomized.py:
```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+):
+    r"""Stochastic Frank-Wolfe (SFW) algorithm.
+
+    The SFW algorithm can solve optimization problems of the form
```
I think this is a bit biased. I think that Mokhtari and other papers in that line of work are more general and can be applied to streams, for example.

GeoffNN replied:
True, I'll note that it is for this implementation.

gideonite commented on copt/randomized.py:
```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+
+    for it in range(max_iter):
+        x_snapshot = x.copy()
+        # Batch size 1
```
I think this should be, `TODO: currently batch size harded to 1`

GeoffNN replied:
I'll commit a change with a `batch_size` parameter, should be straightforward.

gideonite commented on copt/randomized.py:
```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+        step_size = step_size_sfw(variant)
+
+    for it in range(max_iter):
+        x_snapshot = x.copy()
```
`snapshot` is inconsistent with the `_prev` naming convention introduced in line 823

GeoffNN replied:
Agreed, changing to `_prev`

gideonite commented on copt/randomized.py:
```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+        if np.abs(x - x_snapshot).sum() < tol:
+            success = True
+            break
+    message = ""
```
Not familiar with scipy optimize, but are these messages often used? And for what? In this file it seems that it is always empty string, so I guess it's fine.

GeoffNN replied:
Not sure, this was in other stochastic methods from the file that I was using as template. @fabianp ?

gideonite commented on copt/randomized.py:
```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+    assert x.shape == (n_features,)
+    A = sparse.csr_matrix(A).copy()
+
+    dual_var = np.zeros(n_samples)
```
The use of the word dual is ambiguous. I'm increasingly on the side of Lu and Freund who use "prediction vector."

GeoffNN replied:
Here the dual is our alpha, i.e. the f'_i(A_i^Tx). It's the partial gradient, maybe we should call it that?

gideonite replied:
Seems related to the problem of naming `agg` below. Not sure what the best practice is. FWIW I would basically call it whatever but then reference the paper directly in a comment. So either:
* call it dual_var and say `# same as alpha in the paper`
* call it alpha and say `# this represents a stochastic approx. to the true dual representation blah blah see paper for more details`.

Just my two cents on these naming issues.

gideonite commented on copt/randomized.py:
```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+    grad_agg = np.zeros(n_features)
+
+    if variant == 'LF':
+        agg = utils.safe_sparse_dot(A, x)
```
Or is `agg` the prediction vector? Agg = aggregation?

GeoffNN replied:
Yes that's it. `agg` is their `sigma` prediction vector. The other methods (ours + Mokhtari) don't have that variable.

gideonite replied:
See above

gideonite commented on tests/test_stochastic_fw.py:
```diff
@@ -0,0 +1,119 @@
+"""Tests for the Stochastic Frank-Wolfe algorithms."""
```
nitpick: I wish these tests ran faster. I tried superficially to speed them up by making alpha only run over 2 numbers instead of 4. No big change though. I guess running on all four variants each time is slow.

GeoffNN replied:
Sounds good to me!

gideonite replied:
Runs much faster if you reduce the n_samples and n_features. Need to adjust the threshold values accordingly. I can do this if you think it makes sense. Maybe there's some reason I don't see for having higher dimension / more samples in the unit tests.

gideonite commented:
When I pull I have a `test_fw_backtrack(obj, step, alpha)` but this seems to have disappeared here? Anyway, the shape of the trace norm ball is hard coded according to the n_features.
